### PR TITLE
Fix requested received missing txs

### PR DESCRIPTION
**Pull request by SebastianMarian**

### CODE REVIEW

File: process/block/preprocess/basePreProcess.go [hunk 1]
```diff
@@ -218,17 +217,39 @@ func (bpp *basePreProcess) computeExistingAndMissing(
...
+	return bpp.requestMissingTxsForShard(missingTxsForShard, onRequestTxs)
+}
+
+func (bpp *basePreProcess) setMissingTxsForShard(
+	senderShardID uint32,
+	receiverShardID uint32,
+	txHashes [][]byte,
+	forBlock *txsForBlock,
+) {
+	txShardInfoToSet := &txShardInfo{senderShardID: senderShardID, receiverShardID: receiverShardID}
```
Comment (iulianpascalau) [inline process/block/preprocess/basePreProcess.go:236]: multiple lines?

File: process/block/preprocess/basePreProcess.go [hunk 2]
```diff
@@ -218,17 +217,39 @@ func (bpp *basePreProcess) computeExistingAndMissing(
...
+	for _, txHash := range txHashes {
+		forBlock.txHashAndInfo[string(txHash)] = &txInfo{tx: nil, txShardInfo: txShardInfoToSet}
```
Comment (iulianpascalau) [inline process/block/preprocess/basePreProcess.go:238]: multiple lines here?

File: process/block/preprocess/basePreProcess.go [hunk 3]
```diff
@@ -218,17 +217,39 @@ func (bpp *basePreProcess) computeExistingAndMissing(
...
+
 		txHashes = txHashes[:0]
 	}
-	forBlock.mutTxsForBlock.Unlock()
 
-	return missingTxsForShard
+	return bpp.requestMissingTxsForShard(missingTxsForShard, onRequestTxs)
+}
+
+func (bpp *basePreProcess) setMissingTxsForShard(
```
Comment (sasurobert) [inline process/block/preprocess/basePreProcess.go:230]: maybe add comment as it should be called only under mutex

File: process/block/preprocess/basePreProcess.go [hunk 4]
```diff
@@ -218,17 +217,39 @@ func (bpp *basePreProcess) computeExistingAndMissing(
...
+	txHashes [][]byte,
+	forBlock *txsForBlock,
+) {
+	txShardInfoToSet := &txShardInfo{senderShardID: senderShardID, receiverShardID: receiverShardID}
+	for _, txHash := range txHashes {
+		forBlock.txHashAndInfo[string(txHash)] = &txInfo{tx: nil, txShardInfo: txShardInfoToSet}
+	}
+}
+
+func (bpp *basePreProcess) requestMissingTxsForShard(
```
Comment (sasurobert) [inline process/block/preprocess/basePreProcess.go:242]: maybe add comment as it should be called only under mutex

File: process/block/preprocess/basePreProcess.go (hunk 1 above)
Comment (SebastianMarian) [reply process/block/preprocess/basePreProcess.go]: done

File: process/block/preprocess/basePreProcess.go (hunk 2 above)
Comment (SebastianMarian) [reply process/block/preprocess/basePreProcess.go]: done

File: process/block/preprocess/basePreProcess.go (hunk 3 above)
Comment (SebastianMarian) [reply process/block/preprocess/basePreProcess.go:231]: added

File: process/block/preprocess/basePreProcess.go (hunk 4 above)
Comment (SebastianMarian) [reply process/block/preprocess/basePreProcess.go:251]: added

File: process/block/metablock.go [hunk 5]
```diff
@@ -1617,40 +1618,23 @@ func (mp *metaProcessor) requestShardHeaders(metaBlock *block.MetaBlock) (uint32
...
 	for i := 0; i < len(metaBlock.ShardInfo); i++ {
 		shardData := metaBlock.ShardInfo[i]
 		hdr, err := process.GetShardHeaderFromPool(
 			shardData.HeaderHash,
 			mp.dataPool.Headers())
 
 		if err != nil {
-			missingHeadersHashes[shardData.ShardID] = append(missingHeadersHashes[shardData.ShardID], shardData.HeaderHash)
 			mp.hdrsForCurrBlock.missingHdrs++
+			mp.hdrsForCurrBlock.hdrHashAndInfo[string(shardData.HeaderHash)] = &hdrInfo{hdr: nil, usedInBlock: true}
```
Comment (iulianpascalau) [inline process/block/metablock.go:1636]: maybe split on more lines?

File: process/block/shardblock.go [hunk 6]
```diff
@@ -1374,40 +1373,22 @@ func (sp *shardProcessor) requestMetaHeaders(shardHeader *block.Header) (uint32,
...
+
 	for i := 0; i < len(header.MetaBlockHashes); i++ {
 		hdr, err := process.GetMetaHeaderFromPool(
 			header.MetaBlockHashes[i],
 			sp.dataPool.Headers())
 
 		if err != nil {
-			missingHeadersHashes = append(missingHeadersHashes, header.MetaBlockHashes[i])
 			sp.hdrsForCurrBlock.missingHdrs++
+			sp.hdrsForCurrBlock.hdrHashAndInfo[string(header.MetaBlockHashes[i])] = &hdrInfo{hdr: nil, usedInBlock: true}
```
Comment (iulianpascalau) [inline process/block/shardblock.go:1390]: maybe split on multiple lines?

File: process/block/metablock.go (hunk 5 above)
Comment (SebastianMarian) [reply process/block/metablock.go:1636]: done

File: process/block/shardblock.go (hunk 6 above)
Comment (SebastianMarian) [reply process/block/shardblock.go:1390]: done

**Pull request merged**